### ISSUE TITLE
Wrong type check on [numberKey: number]: number

### ISSUE THREAD

**Issue by fthdgn** [rhys-vdw/ts-auto-guard#164]
```ts
/** @see {isNumberNumberMap} ts-auto-guard:type-guard */
export interface NumberNumberMap {
    [numberKey: number]: number
}
```
Generated isNumberNumberMap() function returns false for valid objects. The problem is type of a key is always `string`. However generated code expects `typeof key === "number"`

Generate code should check whether the value is a string representation of a number value.

An example input which typescript compiles:
```ts
let o: NumberNumberMap = {
    1: 101, //or "1"
    1.2: 102, //or "1.2"
    NaN: 103, //or "NaN"
    Infinity: 104, //or "Infinity"
    "-Infinity": 105,
    "-1": 106
}
o[+Infinity] = 107 //or "Infinity" or Infinity
o[-Infinity] = 108 //or "-Infinity"
o[1 / 3] = 109
o[-1] = 110 //or "-1"
```

**Comment by rhys-vdw** [comment]
Hey @fthdgn. Yeah that sounds wrong to me! Just had a quick glance over [this](https://stackoverflow.com/a/175787/317135) and it seems that `!isNaN` might be sufficient. Happy to take a PR to address this.

Do you happen to know whether `number[]` satisfies your interface too? I suspect that it does. Might be worth checking if we handle that case correctly too...

**Comment by fthdgn** [comment]
I didn't know `number[]` can be used as sparse array on Javascript. I will look at it.

But I think `{ key: number]: number }` interface more explanatory than using `number[]` as sparse array. Also my interface is a representation of JSON from server and the structure on the JSON is object.

**Comment by fthdgn** [comment]
I did something like that. But there are many edge cases. I think it would be better to find the official Typescript implementation. 

```ts
function isNumberType(input: string): boolean {
    if (input === "") {
        return false
    } else if (input === "NaN") {
        return true
    } else if (isNaN(input as any)) {
        return false
    } else if (input === "+Infinity") {
        return false
    } else if (input.indexOf("e") > 0 && parseFloat(input) < 1e+21) { //Numbers less than 1e+21 cannot be represented as scientific string
        return false
    } else if (input.match(/\.\d*0+(e.*)?$/)) { //Number cannot be represented with trailing decimal zeroes
        return false
    } else {
        return true
    }
}

function isNumberNumberMap(obj: any, _argumentName?: string): obj is NumberNumberMap {
    return (
        (obj !== null &&
            typeof obj === "object" ||
            typeof obj === "function") &&
        Object.entries(obj)
            .every(([key, value]) => (typeof value === "number" && typeof key === "string" && isNumberType(key)))
    )
}

interface NumberNumberMap {
    [numberKey: number]: number
}

// POSITIVE

let a1: NumberNumberMap = { 1: 1 }
console.log(isNumberNumberMap(a1) === true ? "a1: Pass" : "a1: Error")

let a2: NumberNumberMap = { "1": 1 }
console.log(isNumberNumberMap(a2) === true ? "a2: Pass" : "a2: Error")

let a3: NumberNumberMap = { NaN: 1 }
console.log(isNumberNumberMap(a3) === true ? "a3: Pass" : "a3: Error")

let a4: NumberNumberMap = { Infinity: 1 }
console.log(isNumberNumberMap(a4) === true ? "a4: Pass" : "a4: Error")

let a5: NumberNumberMap = { "-Infinity": 1 }
console.log(isNumberNumberMap(a5) === true ? "a5: Pass" : "a5: Error")

let a6: NumberNumberMap = { "Infinity": 1 }
console.log(isNumberNumberMap(a6) === true ? "a6: Pass" : "a6: Error")

let a8: NumberNumberMap = { "-1": 1 }
console.log(isNumberNumberMap(a8) === true ? "a8: Pass" : "a8: Error")

let a7: NumberNumberMap = {}
a7[1 / 3] = 1
console.log(isNumberNumberMap(a7) === true ? "a7: Pass" : "a11: Error")

let a9: NumberNumberMap = {}
a9[-1] = 1
console.log(isNumberNumberMap(a9) === true ? "a9: Pass" : "a9: Error")

let a10: NumberNumberMap = { 1.2: 1 }
console.log(isNumberNumberMap(a10) === true ? "a10: Pass" : "a10: Error")

let a11: NumberNumberMap = {}
a11[+Infinity] = 1
console.log(isNumberNumberMap(a11) === true ? "a11: Pass" : "a11: Error")

let a12: NumberNumberMap = {}
a12[-Infinity] = 1
console.log(isNumberNumberMap(a12) === true ? "a12: Pass" : "a12: Error")

let a13: NumberNumberMap = { "1e+21": 1 }
console.log(isNumberNumberMap(a13) === true ? "a13: Pass" : "a13: Error")

let a14: NumberNumberMap = { 1e+10: 1 }
console.log(isNumberNumberMap(a14) === true ? "a14: Pass" : "a14: Error")

let a15: NumberNumberMap = { 1e+21: 1 }
console.log(isNumberNumberMap(a15) === true ? "a15: Pass" : "a15: Error")

let a16: NumberNumberMap = { 1_000_000: 1 }
console.log(isNumberNumberMap(a16) === true ? "a16: Pass" : "a16: Error")

let a17: NumberNumberMap = { 1.000: 1 }
console.log(isNumberNumberMap(a17) === true ? "a17: Pass" : "a17: Error")

let a18: NumberNumberMap = { 1.000e+21: 1 }
console.log(isNumberNumberMap(a18) === true ? "a18: Pass" : "a18: Error")

let a19: NumberNumberMap = {}
a19[1e+20 - 1] = 1
console.log(isNumberNumberMap(a19) === true ? "a19: Pass" : "a19: Error")

// NEGATIVE (cast any to allow build)

let n1: NumberNumberMap = {}
n1["+Infinity" as any] = 1
console.log(isNumberNumberMap(n1) === false ? "n1: Pass" : "n1: Error")

let n2: NumberNumberMap = {}
n2["1e+10" as any] = 1
console.log(isNumberNumberMap(n2) === false ? "n2: Pass" : "n2: Error")

let n3: NumberNumberMap = {}
n3["" as any] = 1
console.log(isNumberNumberMap(n3) === false ? "n3: Pass" : "n3: Error")

let n4: NumberNumberMap = {}
n4["1_000_000" as any] = 1
console.log(isNumberNumberMap(n4) === false ? "n4: Pass" : "n4: Error")

let n5: NumberNumberMap = {}
n5["1.000.000" as any] = 1
console.log(isNumberNumberMap(n5) === false ? "n5: Pass" : "n5: Error")

let n6: NumberNumberMap = {}
n6["1.0000" as any] = 1
console.log(isNumberNumberMap(n6) === false ? "n6: Pass" : "n6: Error")

let n7: NumberNumberMap = {}
n7["1.000e+21" as any] = 1
console.log(isNumberNumberMap(n7) === false ? "n7: Pass" : "n7: Error")

```

**Comment by fthdgn** [comment]
I found the typescript code. 
https://raw.githubusercontent.com/microsoft/TypeScript/main/src/compiler/checker.ts line 27009
```ts
        function isNumericLiteralName(name: string | __String) {
            // The intent of numeric names is that
            //     - they are names with text in a numeric form, and that
            //     - setting properties/indexing with them is always equivalent to doing so with the numeric literal 'numLit',
            //         acquired by applying the abstract 'ToNumber' operation on the name's text.
            //
            // The subtlety is in the latter portion, as we cannot reliably say that anything that looks like a numeric literal is a numeric name.
            // In fact, it is the case that the text of the name must be equal to 'ToString(numLit)' for this to hold.
            //
            // Consider the property name '"0xF00D"'. When one indexes with '0xF00D', they are actually indexing with the value of 'ToString(0xF00D)'
            // according to the ECMAScript specification, so it is actually as if the user indexed with the string '"61453"'.
            // Thus, the text of all numeric literals equivalent to '61543' such as '0xF00D', '0xf00D', '0170015', etc. are not valid numeric names
            // because their 'ToString' representation is not equal to their original text.
            // This is motivated by ECMA-262 sections 9.3.1, 9.8.1, 11.1.5, and 11.2.1.
            //
            // Here, we test whether 'ToString(ToNumber(name))' is exactly equal to 'name'.
            // The '+' prefix operator is equivalent here to applying the abstract ToNumber operation.
            // Applying the 'toString()' method on a number gives us the abstract ToString operation on a number.
            //
            // Note that this accepts the values 'Infinity', '-Infinity', and 'NaN', and that this is intentional.
            // This is desired behavior, because when indexing with them as numeric entities, you are indexing
            // with the strings '"Infinity"', '"-Infinity"', and '"NaN"' respectively.
            return (+name).toString() === name;
        }
        ```

**Comment by rhys-vdw** [comment]
> I didn't know number[] can be used as sparse array on Javascript. I will look at it.

Sorry what I meant is... If you're adding this validator for a number-keyed object then it should behave properly when given an array too (perhaps even falling back to the array check if appropriate...).

https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgGIHt3IN4ChkHIDaA1hAJ4BcyIArgLYBG0AutQM5hSgDmuAvrlwJ0ITsgAe1DFgC8xAEQALCABtV6BQBpkCsCqgQFLANy4gA

**Comment by rhys-vdw** [comment]
Closed by #238
